subhub/api/payments.py

```diff
@@ -72,23 +74,23 @@ def cancel_subscription(uid, sub_id) -> FlaskResponse:
         return {"message": "Customer does not exist."}, 404
     customer = stripe.Customer.retrieve(subscription_user.custId)
     for item in customer["subscriptions"]["data"]:
-        if item["id"] == sub_id and item["status"] in ["active", "trialing"]:
+        if item["id"] == sub_id and item["status"] in ["active", "trialing"] and item["cancel_at_period_end"] == False:
             try:
                 tocancel = stripe.Subscription.retrieve(sub_id)
+                stripe.Subscription.modify(sub_id,
+                                           cancel_at_period_end=True)
+                return {"message": "Subscription cancellation successful"}, 201
             except InvalidRequestError as e:
                 # TODO handle other errors: APIConnectionError, APIError, AuthenticationError, CardError
                 return {"message": e}, 400
-            if "No such subscription:" in tocancel:
-                return {"message": "Invalid subscription."}, 404
-            if tocancel["status"] in ["active", "trialing"]:
-                tocancel.delete()
-                return {"message": "Subscription cancellation successful"}, 201
-            else:
-                return {"message": "Error cancelling subscription"}, 400
     else:
         return {"message": "Subscription not available."}, 400
```

Comment:
Is this going to end up returning 'Subscription not available' if the subscription was already cancelled?

Reply:
@bbangert yes it would return 'Subscription not available'.  Is that an issue?  Would changing the message be better or do you want to re-cancel?

Reply:
In the event of a time-out, FxA would not see that it was cancelled successfully and retry. So a repeat call should still work, it just won't need to cancel.

Reply:
OK I changed to this functionality and validated that it will not throw an error.

Reply:
This PR hasn't been updated yet to change that, did you forget to push the change?

Reply:
No, doing a final test and will then push.